pkg/client/restclient/urlprovider.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restclient
+
+import (
+	"net/url"
+	"sync"
+)
+
+// A URLProvider allows to use multiple URLs.
+// Also maintains state of currently working URL.
+type URLProvider interface {
+	// Get returns currently selected URL.
+	Get() *url.URL
+	// Next selects any other URL, it is upto implementation to provide any
+	// health checking/load balancing, it might be as simple as an iterator.
+	Next() *url.URL
+	// WrapWithRetry executes function until success or until all urls were visited.
+	WrapWithRetry(func() error) error
+}
+
+// A RoundRobinProvider allows to iterate over URLs, doesnt provide any health checking.
+type RoundRobinProvider struct {
```

Comment:
> Clients should prefer to connect to the same API Server until it starts
to fail repeatedly

+1

On Thu, Sep 1, 2016 at 1:10 PM, Mike Danese <notifications@github.com>
wrote:

> In pkg/client/restclient/urlprovider.go
> <https://github.com/kubernetes/kubernetes/pull/30588#discussion_r77244373>
> :
>
> > +)
> > +
> > +// A URLProvider allows to use multiple URLs.
> > +// Also maintains state of currently working URL.
> > +type URLProvider interface {
> > +	// Get returns currently selected URL.
> > +	Get() *url.URL
> > +	// Next selects any other URL, it is upto implementation to provide any
> > +	// health checking/load balancing, it might be as simple as an iterator.
> > +	Next() *url.URL
> > +	// WrapWithRetry executes function until success or until all urls were visited.
> > +	WrapWithRetry(func() error) error
> > +}
> > +
> > +// A RoundRobinProvider allows to iterate over URLs, doesnt provide any health checking.
> > +type RoundRobinProvider struct {
>
> Round robin is not an optimal strategy for clients connecting to the
> Kubernetes API Server. Clients should prefer to connect to the same API
> Server until it starts to fail repeatedly. Adding this capability to the
> client is most useful when the clients connect to the API directly. The
> kubernetes service in the default namespace implments this by using service
> SessionAffinity. You could implement this with HA proxy using "balance
> source" and "hash-type consistent". ELBs and GCPLBs support this as well.
> This was also the design of the original provider
> https://github.com/dshulyak/kubernetes/blob/c628b0f279a2581b3b59d43751289c
> f66740c65a/pkg/client/restclient/slightly_sticky_provider.go
>
> This is definitely somehting that we want but I'll let someone else decide
> how important it is. cc @lavalamp <https://github.com/lavalamp>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30588/files/f80530c7f77e2fd272e6718b7d1754df0869d478#r77244373>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngluqyHVih1A1_by4u3iggokMm_30Kks5qlzEagaJpZM4Jj5SF>
> .
>